anndata/utils.py

```diff
@@ -82,11 +82,16 @@ def warn_names_duplicates(attr: str):
     )
 
 
-def warn_no_string_index(names: Sequence[Any]):
+def check_index_is_strings(names: Sequence[Any], attr: str):
+    if isinstance(names, pd.Index) and not isinstance(names.name, str):
+        raise ValueError(
+            f"AnnData expects .{attr}.index.name to be a string, "
+            f"but you passed a name of type {type(names.name).__name__!r}"
+        )
     if not isinstance(names[0], str):
         logger.warning(
-            f"AnnData expects string indices for some functionality, "
-            f"but your first two indices are: {names[:2]}. "
+            f"AnnData expects .{attr}.index to contain strings, "
+            f"but your first indices are: {names[:2]}, …"
         )
```

Comment:
Can we make this an error by now?

Reply:
Don't we automatically convert anyways?

Reply:
Probably worth checking if it'd break anything.

Reply:
`test_set_obs` currently throws that warning, which it shouldn’t. We could do the following to handle this case:

```py
    if not isinstance(names, pd.RangeIndex) and not isinstance(names[0], str):
        raise ValueError(...)
```

I don’t know where we handle a `RangeIndex` being written to disk or used in slicing, but I guess it has always worked and is supposed to work.

Reply:
Looks like you can write it but you can't read it. I think it would be fine for this to be an error, since it generally doesn't work. I don't know if you could even use index using numeric indices.

I just wonder if it'll break something in someone else's code.

Reply:
Hmm, actutally, slicing a RangeIndex will create an Int64Index…

```py
>>> import pandas as pd
>>> pd.DataFrame(dict(a=[1,2,3])).iloc[[0,2], :].index                                                                                                                                                                                   
Out[6]: Int64Index([0, 2], dtype='int64')
```

Reply:
Hmm, actutally, slicing a RangeIndex will create an Int64Index…

```py
>>> import pandas as pd
>>> pd.DataFrame(dict(a=[1,2,3])).iloc[[0,2], :].index                                                                                                                                                                                   
Out[6]: Int64Index([0, 2], dtype='int64')
```

Reply:
I think it will only keep being a range index if you index with `slice`s.

```python
pd.Series(np.ones(10)).iloc[range(4)].index
# RangeIndex(start=0, stop=4, step=1)
```

Interestingly, while you can index a RangeIndex with a range, it results in a Int64Index:

```python
pd.Series(np.ones(10)).iloc[range(4)].index 
# Int64Index([0, 1, 2, 3], dtype='int64')
```

I'm still not quite sure why python has `slice`s and `range`s as separate objects.